### PR TITLE
feat: add `airflow triggerer` Deployment 

### PR DESCRIPTION
**Pull request by mmazek**
## What issues does your PR fix?

- resolves https://github.com/airflow-helm/charts/issues/424

## What does your PR do?

- Adds a new Deployment for the [`airflow triggerer`](https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html#triggerer) so we fully support Airflow's [Deferrable Tasks](https://airflow.apache.org/docs/apache-airflow/stable/concepts/deferring.html) feature.
   - ___WARNING:__ `airflow triggerer` was added in airflow `2.2.0`, so will not work in previous versions_

### For all Pull Requests

- [x] Commits are [signed off](https://github.com/airflow-helm/charts/tree/main/charts/airflow/CONTRIBUTING.md#sign-your-work)
- [x] Commits have [semantic messages](https://github.com/airflow-helm/charts/tree/main/charts/airflow/CONTRIBUTING.md#semantic-commit-messages)
- [x] Documentation [updated](https://github.com/airflow-helm/charts/tree/main/charts/airflow/CONTRIBUTING.md#documentation)
- [x] Passes [ct linting](https://github.com/airflow-helm/charts/tree/main/charts/airflow/CONTRIBUTING.md#linting)

### For releasing ONLY

- [ ] Chart.yaml [version bumped](https://github.com/airflow-helm/charts/tree/main/charts/airflow/CONTRIBUTING.md#versioning)
- [ ] CHANGELOG.md updated

### CODE REVIEW

File: charts/airflow/templates/triggerer/triggerer-deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+            - "-c"
+            - "exec airflow triggerer"
+          {{- if .Values.triggerer.livenessProbe.enabled }}
+          livenessProbe:
+            initialDelaySeconds: {{ .Values.web.livenessProbe.initialDelaySeconds }}
+            periodSeconds: {{ .Values.web.livenessProbe.periodSeconds }}
+            timeoutSeconds: {{ .Values.web.livenessProbe.timeoutSeconds }}
+            failureThreshold: {{ .Values.web.livenessProbe.failureThreshold }}
+            exec:
+              command:
```
Comment (thesuperzapper) [inline charts/airflow/templates/triggerer/triggerer-deployment.yaml:117]: This probe needs to be changed:
- There is no need to set `CONNECTION_CHECK_MAX_COUNT` as we do this in the [_helpers/pods.tpl#L471-L472](https://github.com/airflow-helm/charts/blob/df99a33e3278de9f6935e747e8317794966ed979/charts/airflow/templates/_helpers/pods.tpl#L471-L472)
- The probe command should simply begin with `python`, as the contains already has the entrypoint from  [`include "airflow.command"`](https://github.com/airflow-helm/charts/blob/df99a33e3278de9f6935e747e8317794966ed979/charts/airflow/templates/_helpers/pods.tpl#L15-L22)
- You can use the [liveness probe on the scheduler](https://github.com/airflow-helm/charts/blob/df99a33e3278de9f6935e747e8317794966ed979/charts/airflow/templates/scheduler/scheduler-deployment.yaml#L127-L155) as a reference.

File: charts/airflow/templates/triggerer/triggerer-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                  from airflow.utils.net import get_hostname
+                  import sys
+                  with create_session() as session:
+                      job = session.query(TriggererJob).filter_by(hostname=get_hostname()).order_by(
+                          TriggererJob.latest_heartbeat.desc()).limit(1).first()
+                  sys.exit(0 if job.is_alive() else 1)
+                  "
+          {{- end }}
+          {{- if .Values.web.readinessProbe.enabled }}
+          readinessProbe:
```
Comment (thesuperzapper) [inline charts/airflow/templates/triggerer/triggerer-deployment.yaml:137]: There is no need for a `readinessProbe`, as this deployment is not serving any HTTP services.

File: charts/airflow/values.yaml [hunk 3]
```diff
@@ -857,6 +857,107 @@ workers:
   ##
   extraVolumes: []
 
+###################################
+## COMPONENT | Triggerer
+###################################
+triggerer:
+  ## if the airflow triggerer should be deployed
+  ##
+  enabled: false
```
Comment (thesuperzapper) [inline charts/airflow/values.yaml:866]: There is a question of if the triggerer should be deployed by default, and I am leaning towards yes, but with a warning in the CHANGELOG for users of `2.0.X` and `2.1.X`, and an update to the ["How to use a specific version of airflow?" docs](https://github.com/airflow-helm/charts/tree/main/charts/airflow#how-to-use-a-specific-version-of-airflow).

We don't deploy if `airflow.legacyCommands = true`, so we really only worry about users of `2.0.X` and `2.1.X`.

___NOTE:__ obviously if we add the triggerer, we will make the default version of airflow to be `2.2.X`_

**Review by thesuperzapper — CHANGES_REQUESTED** [review]
Thanks for the PR @mmazek!

I have left some comments inline.

**Comment by mmazek** [comment]
Hi @thesuperzapper, thank you for the feedback! I fixed all the issues you pointed out and also made the triggerer enabled by default.

File: charts/airflow/values.yaml [hunk 4]
```diff
@@ -857,6 +857,98 @@ workers:
   ##
   extraVolumes: []
 
+###################################
+## COMPONENT | Triggerer
+###################################
+triggerer:
+  ## if the airflow triggerer should be deployed
```
Comment (thesuperzapper) [inline charts/airflow/values.yaml:864]: Update docstring with some warnings:
```yaml
triggerer:
  ## if the airflow triggerer should be deployed
  ## - [WARNING] the triggerer component was added in airflow 2.2.0
  ## - [WARNING] if `airflow.legacyCommands` is `true` the triggerer will NOT be deployed
  ##
  enabled: true
```

File: charts/airflow/README.md [hunk 5]
```diff
@@ -1556,6 +1556,33 @@ Parameter | Description | Default
...
 </details>
 
+### `triggerer.*`
+<details>
+<summary>Expand</summary>
+<hr>
+
+Parameter | Description                                                                             | Default
+--- |-----------------------------------------------------------------------------------------| ---
+`triggerer.enabled` | if the triggerer should be deployed. `airflow.legacyCommands` should be `false` as well | `false`
```
Comment (thesuperzapper) [inline charts/airflow/README.md:1566]: The default value is now `true`, and description is always the same as the first line of the docstring, so in this case:
```md
`triggerer.enabled` | if the triggerer should be deployed | `true`
```

File: charts/airflow/README.md [hunk 6]
```diff
@@ -1556,6 +1556,33 @@ Parameter | Description | Default
 <hr>
 </details>
 
+### `triggerer.*`
```
Comment (thesuperzapper) [inline charts/airflow/README.md:1559]: _(not related to this section)_

Can you please also add the triggerer to the ["How to use Pod affinity, nodeSelector, and tolerations?"](https://github.com/airflow-helm/charts/blob/df99a33e3278de9f6935e747e8317794966ed979/charts/airflow/README.md?plain=1#L1374-L1378) section of the FAQ:

```yaml
## airflow triggerer
triggerer:
  nodeSelector: {}
  affinity: {}
  tolerations: []
```

File: charts/airflow/values.yaml [hunk 7]
```diff
@@ -857,6 +857,98 @@ workers:
...
+  ##   https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#volumemount-v1-core
+  ##
+  extraVolumeMounts: []
+
+  ## extra Volumes for the triggerer Pods
+  ## - spec for Volume:
+  ##   https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#volume-v1-core
+  ##
+  extraVolumes: []
+
```
Comment (thesuperzapper) [inline charts/airflow/values.yaml:942]: Can you add a new value called `triggerer.capacity` to set the [`airflow triggerer --capacity`](https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html#triggerer) flag:
```yaml
  ## sets the  `--capacity` parameter used to run the airflow triggerer
  ##
  capacity: 1000
```

NOTE: in terms of ordering, it should go above the `livenessProbe` values

**Review by thesuperzapper — CHANGES_REQUESTED** [review]
Hey @mmazek great work so far, but still a few quick changes are required to get it merged.

File: charts/airflow/values.yaml [hunk 8]
```diff
@@ -857,6 +857,120 @@ workers:
...
+  ## Pod annotations for the triggerer Deployment
+  ##
+  podAnnotations: {}
+
+  ## if we add the annotation: "cluster-autoscaler.kubernetes.io/safe-to-evict" = "true"
+  ##
+  safeToEvict: true
+
+
```
Comment (thesuperzapper) [inline charts/airflow/values.yaml:925]: There is a random extra newline here.

File: charts/airflow/templates/triggerer/triggerer-deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+          envFrom:
+            {{- include "airflow.envFrom" . | indent 12 }}
+          env:
+            {{- include "airflow.env" . | indent 12 }}
+          command:
+            {{- include "airflow.command" . | indent 12 }}
+          args:
+            - "bash"
+            - "-c"
+            - "exec airflow triggerer --capacity {{ .Values.triggerer.capacity }}"
```
Comment (thesuperzapper) [inline charts/airflow/templates/triggerer/triggerer-deployment.yaml:108]: We can instead have the `triggerer.capacity` value set the `AIRFLOW__TRIGGERER__DEFAULT_CAPACITY` config in [./config/secret-config-envs.yaml#L113](https://github.com/airflow-helm/charts/blob/df99a33e3278de9f6935e747e8317794966ed979/charts/airflow/templates/config/secret-config-envs.yaml) (after line 113):

```yaml
  {{- if and .Values.triggerer.enabled (not .Values.airflow.legacyCommands) }}
  AIRFLOW__TRIGGERER__DEFAULT_CAPACITY: {{ .Values.triggerer.capacity | toString | b64enc | quote }}
  {{- end }}
```

File: charts/airflow/values.yaml [hunk 10]
```diff
@@ -857,6 +857,120 @@ workers:
...
+    ## the maximum unavailable pods/percentage for the triggerer Deployment
+    ##
+    maxUnavailable: ""
+
+    ## the minimum available pods/percentage for the triggerer Deployment
+    ##
+    minAvailable: ""
+
+  ## sets the  `--capacity` parameter used to run the airflow triggerer
```
Comment (thesuperzapper) [inline charts/airflow/values.yaml:941]: We will also need to update the docstring to reference `AIRFLOW__TRIGGERER__DEFAULT_CAPACITY` is what is being set.

```
## maximum number of triggers each triggerer will run at once (sets `AIRFLOW__TRIGGERER__DEFAULT_CAPACITY`)
```

PS: also don't forget to update the docstring in the readme.

**Review by thesuperzapper — CHANGES_REQUESTED** [review]
Upon testing I realized that the `--capacity` parameter actually does not work (its airflow's fault, see https://github.com/apache/airflow/issues/21752), setting it at all causes the triggerer to crash.

File: charts/airflow/README.md [hunk 11]
```diff
@@ -1556,6 +1562,35 @@ Parameter | Description | Default
...
+`triggerer.annotations` | annotations for the triggerer Deployment                                            | `{}`
+`triggerer.podAnnotations` | Pod annotations for the triggerer Deployment                                        | `{}`
+`triggerer.safeToEvict` | if we add the annotation: "cluster-autoscaler.kubernetes.io/safe-to-evict" = "true" | `true`
+`triggerer.podDisruptionBudget.*` | configs for the PodDisruptionBudget of the triggerer Deployment                     | `<see values.yaml>`
+`triggerer.capacity` | maximum number of triggers each triggerer will run at once (sets `AIRFLOW__TRIGGERER__DEFAULT_CAPACITY`)                  | 1000
+`triggerer.livenessProbe.*` | liveness probe for the triggerer Pods                                               | `<see values.yaml>`
+`triggerer.extraPipPackages` | extra pip packages to install in the triggerer Pods                                 | `[]`
+`triggerer.extraVolumeMounts` | extra VolumeMounts for the triggerer Pods                                           | `[]`
+`triggerer.extraVolumes` | extra Volumes for the triggerer Pods                                                | `[]]`
```
Comment (thesuperzapper) [inline charts/airflow/README.md:1589]: This is quite messy, can we make it consistent with the other tables:

```md
Parameter | Description | Default
--- | --- | ---
`triggerer.enabled` | if the triggerer should be deployed | `true`
`triggerer.replicas` | the number of triggerer Pods to run | `1`
`triggerer.resources` | resource requests/limits for the airflow triggerer Pods | `{}`
`triggerer.nodeSelector` | the nodeSelector configs for the triggerer Pods | `{}`
`triggerer.affinity` | the affinity configs for the triggerer Pods | `{}`
`triggerer.tolerations` | the toleration configs for the triggerer Pods | `[]`
`triggerer.securityContext` | the security context for the triggerer Pods | `{}`
`triggerer.labels` | labels for the triggerer Deployment | `{}`
`triggerer.podLabels` | Pod labels for the triggerer Deployment | `{}`
`triggerer.annotations` | annotations for the triggerer Deployment | `{}`
`triggerer.podAnnotations` | Pod annotations for the triggerer Deployment | `{}`
`triggerer.safeToEvict` | if we add the annotation: "cluster-autoscaler.kubernetes.io/safe-to-evict" = "true" | `true`
`triggerer.podDisruptionBudget.*` | configs for the PodDisruptionBudget of the triggerer Deployment | `<see values.yaml>`
`triggerer.capacity` | maximum number of triggers each triggerer will run at once (sets `AIRFLOW__TRIGGERER__DEFAULT_CAPACITY`) | `1000`
`triggerer.livenessProbe.*` | liveness probe for the triggerer Pods | `<see values.yaml>`
`triggerer.extraPipPackages` | extra pip packages to install in the triggerer Pods | `[]`
`triggerer.extraVolumeMounts` | extra VolumeMounts for the triggerer Pods | `[]`
`triggerer.extraVolumes` | extra Volumes for the triggerer Pods | `[]`
```

**Review by thesuperzapper — CHANGES_REQUESTED** [review]
@mmazek looks good now in terms of the triggerer itself, but still one strange formatting on the README parameter table.

___PS:__ I would have pushed the small change to your branch myself, but I think because you are using the `main` branch of your fork, it is not letting me push commits to your fork. Or you might have just not ticked `Allow edits and access to secrets by maintainers` on the PR sidebar._

File: charts/airflow/README.md (hunk 11 above)
Comment (mmazek) [reply charts/airflow/README.md]: Sorry, it was PyCharm Markdown plugin acting weird. It's fixed now.

**Review by thesuperzapper — APPROVED** [review]
@mmazek looks good, thanks for all your work and patience getting this ready!

I will merge when I cut the `8.6.0` release (hopefully soon, my goal is to make `8.6.0` the "prod-ready Airflow `2.2.X`" release)

**Comment by thesuperzapper** [comment]
@mmazek in the interest of releasing `8.6.0` ASAP, I have made a new PR https://github.com/airflow-helm/charts/pull/555.

This PR includes your commit (so you should end up with the "contributor" tag).

I had to do this because you did not allow maintainers of the upstream repo to add commits when you created your PR.

**Pull request closed without merging**